### ISSUE TITLE
Need an unprotected Peer API to retrieve security settings

### ISSUE THREAD

**Issue by smithbk** [hyperledger-archives/fabric#1276]
In order to make the client SDK easier to use, we need a GRPC Peer API to return the chain's security settings.  The client SDK will then be able to accept or reject communicating with the peer/chain.

**Comment by corecode** [comment]
Shouldn't the client be configured for a specific chain, i.e. with appropriate TLS certificates, if security is enabled?

**Comment by adecaro** [comment]
Having an unprotected Peer API here could lead to downgrade security attacks. A man in the middle could potentially force a client-sdk to use lower security levels.

**Comment by smithbk** [comment]
I'm thinking of a similar approach used by TLS today in terms of negotiating which cipher suite to use.  Is that open to a man-in-the-middle attack?

**Comment by adecaro** [comment]
@smithbk, it is all about how you design the protocol and how you implement it. Openssl was prone to such attacks (https://en.wikipedia.org/wiki/Downgrade_attack), for example.

**Comment by JonathanLevi** [comment]
Yes, @smithbk I like TLS too. It is probably reasonable to adopt a standard approach, such as the one used by TLS v1.2 and making sure that the right algorithms/key-lengths combinations are used. 

Keeping it practical, if this logic is implemented correctly [which is probably one of @adecaro's concerns above, understandably] it should do.

From the use-cases I see, cipher hardening will be key here... sometimes at the price of some non-zero initial investment (set-up) configurations, yes.

**Comment by JonathanLevi** [comment]
@corecode Yes, I think so. It will be easier to pre-configure (a few) trusted authorities [their public keys] from which one requests [E/T]Certs... so that such certs could be validated/verified by the client (using the signer's public key).

**Comment by mastersingh24** [comment]
Maybe I am missing something here, but let's assume that security is enabled for the chain.  In this case, a client will need to register with membership services and will receive / retrieve all of the relevant enrollment material need to access the chain.  It should be quite forward to have the peer expose a secure API for obtaining any relevant bootstrap settings:
- require the API call to be signed (just like normal transactions).  We can require that the client specify the signature and hash algorithm being used (as long as it is one supported by the peer and there will be a finite set supported) or we can be fancy and infer this info from the SubjectPublicKey info in the public cert presented - e.g. ECC curve p384 and this would imply SHA3-384, etc
- if we require encryption for this, again we could spec this out (ECDH plus AES256), etc

So there is no need for this to be totally "open"

**Comment by angrbrd** [comment]
It's not clear to me if this is something that still needs to be addressed? And if we have agreed on the way this needs to be addressed? Please let me know. Thanks!

**Comment by smithbk** [comment]
We still need an API to get the security settings, though as Gari said, it doesn't have to be unprotected as the title of this item says.

**Comment by angrbrd** [comment]
I have created a matching issue in Jira here: https://jira.hyperledger.org/browse/FAB-124

@christo4ferris please close out this issue. Thanks!
